### ISSUE TITLE
5.10 is buggy with linked nodes in custom schema. They are null until cache clean and rebuild. Will eventually corrupt again.

### ISSUE THREAD

**Issue by tony-gutierrez** [gatsbyjs/gatsby#38203]
### Preliminary Checks

- [X] This issue is not a duplicate. Before opening a new issue, please search existing issues: https://github.com/gatsbyjs/gatsby/issues
- [X] This issue is not a question, feature request, RFC, or anything other than a bug report directly related to Gatsby. Please post those things in GitHub Discussions: https://github.com/gatsbyjs/gatsby/discussions

### Description

Linked nodes in custom schema are no longer being reliably populated. Corruption is being cached. Nodes return null.

### Reproduction Link

https://githubbox.com/gatsbyjs/gatsby-starter-minimal

### Steps to Reproduce

1. Create a working site, with a custom schema
```type ContentfulImageAssets implements ContentfulReference & ContentfulEntry & Node {
            fullUrl: String
            asset: ContentfulAsset
            @link(from: "asset___NODE")
            brandfolderAsset: [contentfulImageAssetsBrandfolderAssetJsonNode]
            @link(from: "brandfolderAsset___NODE")
        }
```
2. Do incremental builds in Gatsby cloud by updating content
3. Evenutally "asset" will return null
4. Clear cache and force a manual rebuild in Gatsby cloud, and your content is back
...

### Expected Result

Reliability

### Actual Result

Missing fields

### Environment

```shell
System:
    OS: Linux 5.15 Ubuntu 20.04.4 LTS (Focal Fossa)
    CPU: (24) x64 AMD Ryzen 9 5900X 12-Core Processor
    Shell: 5.0.17 - /bin/bash
  Binaries:
    Node: 18.16.0 - ~/.nvm/versions/node/v18.16.0/bin/node
    npm: 9.5.1 - ~/.nvm/versions/node/v18.16.0/bin/npm
  npmPackages:
    gatsby: ^5.10.0 => 5.10.0 
    gatsby-plugin-canonical-urls: ^5.10.0 => 5.10.0 
    gatsby-plugin-gatsby-cloud: ^5.10.0 => 5.10.0 
    gatsby-plugin-image: ^3.10.0 => 3.10.0 
    gatsby-plugin-node-fields: ^3.1.0 => 3.1.0 
    gatsby-plugin-react-helmet: ^6.10.0 => 6.10.0 
    gatsby-plugin-sitemap: ^6.10.0 => 6.10.0 
    gatsby-source-contentful: ^8.10.0 => 8.10.0 
  npmGlobalPackages:
    gatsby: 5.10.0
```

### Config Flags

DEV_SSR: true

**Comment by LekoArts** [comment]
Hi!

Sorry to hear you're running into an issue. To help us best begin debugging the underlying cause, it is incredibly helpful if you're able to create a [minimal reproduction][reproduction]. This is a simplified example of the issue that makes it clear and obvious what the issue is and how we can begin to debug it.

If you're up for it, we'd very much appreciate if you could [provide a minimal reproduction][reproduction] and we'll be able to take another look.

Thanks for using Gatsby! 💜

[reproduction]: https://gatsby.dev/reproduction

**Comment by tony-gutierrez** [comment]
When the cache is corrupted I will see things like this during build:
```
warn Multiple node fields resolve to the same GraphQL field `ContentfulImageAssets.asset` - [`asset___NODE`, `asset`]. Gatsby will use `asset___NODE`.
```

**Comment by tony-gutierrez** [comment]
Downgraded to 5.9 and get the same issue occasionally. It's like the schema isn't consistent despite the source being consistent.

I am not sure if the problem is in gatsby-source-contentful or gatsby.

```
Multiple node fields resolve to the same GraphQL field contentfulPerspectiveSummaryTextNode.summary - [summary, summary___NODE]. Gatsby will use summary___NODE.

Multiple node fields resolve to the same GraphQL field contentfulBrochureSummaryTextNode.summary - [summary, summary___NODE]. Gatsby will use summary___NODE.
```

I am having to clear cache and rebuild production multiple times a day.

**Comment by laurent-le-graverend** [comment]
Can't be happier to find this issue, haha.
It's been days we are struggling with this too.

Sometimes Gatsby builds are ok (on Gatsby Cloud or locally).
Often the cache seems to get corrupted and only a manual "clear cache and restart" can solve it (it may still fail).

Details:
```
Node version v18.16.1
Gatsby version Gatsby CLI version: 5.11.0
Gatsby version: 5.11.0
gatsby-source-contentful: 8.11.0```

Curious things that are happening in a failed build:

```
verbose 55.671506293 Plugin "gatsby-source-contentful" is using the ___NODE convention which is deprecated. This plugin should use the @link directive instead.

warning Multiple node fields resolve to the same GraphQL field `contentfulBrandDescriptionTextNode.description` - [`description`, `description___NODE`]. Gatsby will use `description___NODE`.

Migration: https://gatsby.dev/node-convention-deprecation

verbose 55.699907411 Plugin "null" is using the ___NODE convention which is deprecated. This plugin should use the @link directive instead.
```

Note the plugin "null", that's not appearing on a successful build.

Then it crashes on this error.

```
ERROR There was an error in your GraphQL query:

Field "description" of type "contentfulBrandDescriptionTextNode" must have a selection of subfields. Did you mean "description { ... }"?
```

In short, the graphQL schema seems to be corrupted, and the related fields are not the right match anymore.

**Comment by tony-gutierrez** [comment]
I get that same problem sometimes...it thinks contentful long text fields should have an additional subfield, until you clear the cache. The other problem is the null for linked nodes...until you clear the cache. Gatsby is moving too fast on the core, without bringing the rest along with it.

**Comment by shameer-rahman** [comment]
facing the same issue on gatsby v5.10, we're frustrated with this error where we need to clear cache and build.

**Comment by tony-gutierrez** [comment]
@LekoArts Can we get the team to look into this? We cant exactly recreate in in a minimal repo, but it seems pretty widespread.

**Comment by tony-gutierrez** [comment]
> facing the same issue on gatsby v5.10, we're frustrated with this error where we need to clear cache and build.

@shameer-rahman What source plugin are you using? Contentful?

**Comment by LekoArts** [comment]
We can look into it if we have something we _actually_ can look into as unfortunately no project or reproduction was shared yet. You can also read https://antfu.me/posts/why-reproductions-are-required if that makes more sense.

If this only happens on Gatsby Cloud, please contact the support there. If this also happens locally with `gatsby build`, please provide a minimal reproduction as requested.

**Comment by shameer-rahman** [comment]
> > facing the same issue on gatsby v5.10, we're frustrated with this error where we need to clear cache and build.
> 
> @shameer-rahman What source plugin are you using? Contentful?

yes, we're using contentful, we're building our website using azure self hosted agent where we usually encounter this issue.

**Comment by tony-gutierrez** [comment]
> We can look into it if we have something we _actually_ can look into as unfortunately no project or reproduction was shared yet. You can also read https://antfu.me/posts/why-reproductions-are-required if that makes more sense.
> 
> If this only happens on Gatsby Cloud, please contact the support there. If this also happens locally with `gatsby build`, please provide a minimal reproduction as requested.

@LekoArts Not all bugs are easy to recreate. Sometimes you just have to take the clues you are given and go with them. The bug does happen in local builds, but we would need a dummy Contentful environment to start with a recreation. I would think the fact that the Contenful source plugin is throwing a deprecated error would be enough to start from. You also have the concept of the schema creating more than one field, and not knowing which to use: `[summary, summary___NODE]`.

**Comment by laurent-le-graverend** [comment]
Sounds like a good summary of what is happening here @tony-gutierrez 👍

**Comment by LekoArts** [comment]
> but we would need a dummy Contentful environment to start with a recreation

Then please do that, otherwise we can't help you

**Comment by axe312ger** [comment]
Sorry, I just came across this now. @LekoArts feel free to mention me on Contentful tickets! :)

@tony-gutierrez & @laurent-le-graverend - these bugs are unfortunately very hard to reproduce, can you please answer me these questions, so I can narrow it down?

1. Did you see similar issues with earlier versions of the plugin?
2. How much editor traffic do you have between builds? One edit? None? Many edits? Is one person doing the edits or is it a team of people?
3. Do you use the preview feature of Gatsby & Contentful?
4. If yes, does this happen only for preview builds, or also on live/master/production builds?

Thanks!

**Comment by tony-gutierrez** [comment]
We have schema customization, but it is really only there to prevent failed builds when the last node of a type gets deleted. Looks like this in gatsby_node.js:
```
type ContentfulBrochure implements ContentfulReference & ContentfulEntry & Node {
            components: [AllComponents]
            @link(from: "components___NODE")
            seoAndMetadata: ContentfulSeoMetadata
            @link(from: "seoAndMetadata___NODE")
            hideOnInsightsPage: Boolean
            customBackgroundImage: ContentfulImageAssets
            @link(from: "customBackgroundImage___NODE")
        }
 ```

It doesn't seem to line up directly with the random schema errors, and has been around for a long time with no issues.

**Comment by axe312ger** [comment]
The depreciation warning is not an issue. Everyone has this using the current version of the Contentful plugin.

We have a new version coming up, it will generate the schema in the proper way Gatsby expects it. 

https://github.com/gatsbyjs/gatsby/issues/31385
https://github.com/gatsbyjs/gatsby/pull/30855

**Comment by tony-gutierrez** [comment]
@axe312ger So is the plan to just leave the bug in place until the next version is finalized? I am not positive that the contentful source plugin is the cause of the issue, but it is being experienced by paying users....

**Comment by laurent-le-graverend** [comment]
@axe312ger do you have an ETA for the next version please?
It's a tough issue to handle as it breaks Gatsby Cloud builds randomly almost everyday.

**Comment by tony-gutierrez** [comment]
I downgraded and I haven't seen the error in a while. I believe I'm using
8.6.0.

**Comment by shameer-rahman** [comment]
The error still continues even after downgrading  gatsby-source-contentful to 8.8.0, i also tried 8.6.0, in both the versions the build is getting corrupt eventually. @tony-gutierrez is it the same with you?

**Comment by axe312ger** [comment]
The patch is public via `gatsby-source-contentful@8.13.0-next.1`

**Comment by Swissbite** [comment]
I'm happy to find this bug report here. @axe312ger I can confirm that the version `gatsby-source-contentful@8.13.0-next.1` solves the issue. Is there a date for the release of `gatsby-source-contentful@8.13.x`?

By the way: The statistics output now shows micro-updates? Or how can I interpret these decimal places?
```
success Contentful: Fetch data - 0.592s
info Contentful: 1 new entries
info Contentful: 1 updated entries
info Contentful: 1 deleted entries
info Contentful: 1022.3333333333334 cached entries
info Contentful: 0 new assets
info Contentful: 0 updated assets
info Contentful: 734.6666666666666 cached assets
info Contentful: 0 deleted assets
```

**Comment by axe312ger** [comment]
> I'm happy to find this bug report here. @axe312ger I can confirm that the version `gatsby-source-contentful@8.13.0-next.1` solves the issue. Is there a date for the release of `gatsby-source-contentful@8.13.x`?

Great its solved for you as well. It will be released with the next Gatsby minor update. So pretty soon :)

> 
> By the way: The statistics output now shows micro-updates? Or how can I interpret these decimal places?
> 
> ```
> success Contentful: Fetch data - 0.592s
> info Contentful: 1 new entries
> info Contentful: 1 updated entries
> info Contentful: 1 deleted entries
> info Contentful: 1022.3333333333334 cached entries
> info Contentful: 0 new assets
> info Contentful: 0 updated assets
> info Contentful: 734.6666666666666 cached assets
> info Contentful: 0 deleted assets
> ```

This whole stats and progress are a little bit off, we greatly improved it with the next major version of the plugin. See #31385

**Comment by dslovinsky** [comment]
We upgraded to `8.13.0-next.1` but are still experiencing this problem. It's usually a Contentful long text field that stops being recognized by graphQL after a seemingly random number of cached builds. Clearing the cache always fixes it. Seeing the same issue on three projects total.

```
There was an error in your GraphQL query:

Field "bio" of type "contentfulEntityPersonBioTextNode" must have a selection of subfields. Did you mean "bio { ... }"?
```
<img width="763" alt="image" src="https://github.com/gatsbyjs/gatsby/assets/65476034/30b9bd1e-2268-4793-a5e4-7ae9a3c1fd10">

**Comment by ZempTime** [comment]
Have a similar error report with some additional info in case that would be helpful.

We're seeing the same as https://github.com/gatsbyjs/gatsby/issues/38203#issuecomment-1718484063 -

```
11:06:20 AM: error There was an error in your GraphQL query:
11:06:20 AM: Field description of type contentfulBlogCategoryDescriptionTextNode must have a selection of subfields. Did you mean description { ... }?
```

And additionally, we will get this error on other builds which seems related to the same type:

```
11:02:05 AM: error There was an error in your GraphQL query:
11:02:05 AM: Field "ne" is not defined by type "contentfulBlogPostDescriptionTextNodeFilterInput".
11:02:05 AM:
11:02:05 AM:
11:02:05 AM:   GraphQLError: Field "ne" is not defined by type "contentfulBlogPostDescription  TextNodeFilterInput".
```

Clearing cache & rebuilding results in a successful build.

**Comment by Swissbite** [comment]
@dslovinsky Does the release https://github.com/gatsbyjs/gatsby/releases/tag/gatsby-source-contentful@8.12.1 include this fix?

**Comment by laurent-le-graverend** [comment]
No releases notes apparently. Let's find out 😅 

I've noticed something; it seems to happen only when an image is uploaded on a Contentful item.

**Comment by axe312ger** [comment]
> No releases notes apparently. Let's find out 😅
> 
> I've noticed something; it seems to happen only when an image is uploaded on a Contentful item.

If its not in the changelog, its very unlikely its in the release. 

The fix is in the beta of 8.13, which should cause no issues installing on top of any 8.x installation of the plugin.

Gatsby releases are a little slow at the moment, feel free to use the 8.13 beta. (`npm i gatsby-source-contentful@8.13.0-next.1`)


We also run a beta test of the next major version of the plugin. As it removes a lot of "background magic", the new major should fix many issues releated to data model, empty fields, broken references and so on:

https://github.com/gatsbyjs/gatsby/discussions/38585
